src/main/java/io/temporal/internal/statemachines/ChildWorkflowStateMachine.java

```diff
@@ -0,0 +1,305 @@
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.statemachines;
+
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.StartChildWorkflowExecutionCommandAttributes;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.api.common.v1.WorkflowExecution;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.enums.v1.EventType;
+import io.temporal.api.enums.v1.RetryState;
+import io.temporal.api.enums.v1.TimeoutType;
+import io.temporal.api.history.v1.ChildWorkflowExecutionCanceledEventAttributes;
+import io.temporal.api.history.v1.ChildWorkflowExecutionCompletedEventAttributes;
+import io.temporal.api.history.v1.ChildWorkflowExecutionFailedEventAttributes;
+import io.temporal.api.history.v1.ChildWorkflowExecutionTerminatedEventAttributes;
+import io.temporal.api.history.v1.ChildWorkflowExecutionTimedOutEventAttributes;
+import io.temporal.api.history.v1.StartChildWorkflowExecutionFailedEventAttributes;
+import io.temporal.client.WorkflowExecutionAlreadyStarted;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.failure.CanceledFailure;
+import io.temporal.failure.ChildWorkflowFailure;
+import io.temporal.failure.TerminatedFailure;
+import io.temporal.failure.TimeoutFailure;
+import io.temporal.internal.replay.ChildWorkflowTaskFailedException;
+import io.temporal.workflow.Functions;
+import java.util.Optional;
+
+final class ChildWorkflowStateMachine
+    extends EntityStateMachineInitialCommand<
+        ChildWorkflowStateMachine.State,
+        ChildWorkflowStateMachine.ExplicitEvent,
+        ChildWorkflowStateMachine> {
+
+  enum ExplicitEvent {
+    SCHEDULE,
+    CANCEL
+  }
+
+  enum State {
+    CREATED,
+    START_COMMAND_CREATED,
+    START_EVENT_RECORDED,
+    STARTED,
+    START_FAILED,
+    COMPLETED,
+    FAILED,
+    CANCELED,
+    TIMED_OUT,
+    TERMINATED,
+  }
+
+  public static final StateMachineDefinition<State, ExplicitEvent, ChildWorkflowStateMachine>
+      STATE_MACHINE_DEFINITION =
+          StateMachineDefinition.<State, ExplicitEvent, ChildWorkflowStateMachine>newInstance(
+                  "ChildWorkflow",
+                  State.CREATED,
+                  State.START_FAILED,
+                  State.COMPLETED,
+                  State.FAILED,
+                  State.CANCELED,
+                  State.TIMED_OUT,
+                  State.TERMINATED)
+              .add(
+                  State.CREATED,
+                  ExplicitEvent.SCHEDULE,
+                  State.START_COMMAND_CREATED,
+                  ChildWorkflowStateMachine::createStartChildCommand)
+              .add(
+                  State.START_COMMAND_CREATED,
+                  CommandType.COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION,
```

Comment:
What are the side effects of not defining the no-op transitions that are command based?

Reply:
Only defined transitions are allowed. Workflow task fails if command or event is applied to the state that doesn't have a transition defined for it.